Rewrites HandleSummary() doc (#842)
grafana/k6-docs: src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md

---
title: Custom summary
excerpt: With the handlesummary() function, you can customize every part of your report. Change the content, redirect output, and more.
---

> With the `handleSummary()` callback, you can completely customize your end-of-test summary.

Author: ```suggestion
With the `handleSummary()` callback, you can completely customize your end-of-test summary.
```

Is there any reason to specify that it's a "callback" function? Or is this exposing too much internals? What function does it call back to?

If it's necessary to mention that it is a callback, maybe we could add an explanatory sentence at the end of the "About handleSummary()) section.

Reviewer: I am for Less is More. So, in this context, it could be unnecessary. 

> With `handleSummary()`, you can 

[Read more about callbacks](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Introducing#callbacks)

Reviewer 2: I would argue it shouldn't be called callback function.

From the [MDN article](https://developer.mozilla.org/en-US/docs/Glossary/Callback_function) the definition is 
> a function passed into another function as an argument

We are missing the whole "passed into another function" so IMO we can drop it.

If we don't I would argue we should call `setup`, `teardown` and `default` callback functions as well ;)

Reviewer: I'd separate this convo in another thread. Here my thoughts:

1 - I think `default` would be different --> the default scenario function?

2 - But I agree `setup`, `teardown`, and `handleSummary` follow the same pattern.

3.1 - I have seen using `callback` for this case. But it is confusing.
3.2 - Another common term in JS land is [hook](https://mochajs.org/#hooks)
3.3 - To keep it simple - I am leaning toward using the term `global functions`?

Author: Yes, this is beyond the scope of this PR, but I have two questions:

These functions are call backs to what other function? So it seems like these functions  run within a "super-function" that k6 automatically sets up when the script runs?

Is this detailed information about runtime useful to any users? If it is, we could probably add it to the Test life cycle document.

Reviewer 2: > these functions run within a "super-function" that k6 automatically sets up when the script runs?

This won't be how I will explain it. K6 just happens to be calling this functions in a particular pattern that can be explained in the Test life cycle.

> I am leaning toward using the term global functions?

I am for "exported function"  as it (and the rest) need to be *exported* and a *function* so it seems like the best choice to me.

Author: Sorry, just for my understanding, isn't there some kind of "superfunction" under the hood? I mean, _to what_ are these functions exported?

I know this is way too technical, and I beg forgiveness for my poor vocabulary, but it has confused me for a while. I think it's related to how I got so confused by "where" the data in `handlesummary(data)` came from.

Reviewer: I think the term "exported function" lacks some more context as any k6/JS module can export a function. 

Moving the convo to https://github.com/grafana/k6-docs/issues/845

Reviewer 2: > Sorry, just for my understanding, isn't there some kind of "superfunction" under the hood? I mean, to what are these functions exported?

They are exported as [defined](https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export) by the ecmascript specification. Which means that other JS modules can use them. 

k6 is not JS code it just uses the same mechanic to call them. 

I guess you can call it a superfunction (there is no such term, and it is not a [higher-order function](https://en.wikipedia.org/wiki/Higher-order_function)) it is just that k6 calls those functions with given parameters in a given ... pattern. They are [called](https://github.com/grafana/k6/blob/8a74171c8b43f48fb53078e0bcfa7851828124bc/js/runner.go#L321) from things that *are* functions within the go language :shrug:. 

But I am not really certain how this is relevant to anyone :shrug: 

>  got so confused by "where" the data in handlesummary(data) came from.

It comes from k6 providing it. 

In practice *almost* everything in computers comes from another program. When you run your program your OS provides it with a bunch of stuff in order to do it. In the same way we provide the info to the script when we execute it. It is just that OSes have some "standards" on how they will execute a program and we have some "standards" on how we will execute the script. And for the most parts most of the users do not care about *most* of those. 

In the cases of the functions we are talking about, you need to know that the first argument will be the data returned from setup or in the case of handleSummary the "summary data". Both of those are just things that when we designed the system decided will be useful. 

To be honest if we make this today - there is a good chance that some of those will just be things `k6/execution` will be able to return to you instead. As documenting that seems a lot easier and is definitely less confusing. Or at least we would have a discussion about it :shrug: 

Hope this helps you!

In this document, you can read about:
- How `handleSummary()` works
- How to customize the content and output location of your summary
- The data structure of the summary object

<Blockquote mod="note"
title="handleSummary() is available only for local tests.">

However, we plan to support the feature for k6 Cloud tests, too.
[Track progress in this issue](https://github.com/grafana/k6-cloud-feature-requests/issues/24).

</Blockquote>

## About the `handleSummary()` callback

> After your VU code runs, k6 aggregates your metrics into an object.

Reviewer: We mostly talk about test results and end-of-test summary. The word `object` does not sound familiar at this moment. 

One attempt:

> After k6 executes the load test, it aggregates the test results into a summary. The handleSummary() function receives this summary as an argument (called data in all examples here). 

> You can use handleSummary() to manipulate the test result summary as you want. To get an idea of what the data looks like, run this script and open the output file, summary.json.

Reviewer 2: I am also against the wording of

> VU code runs

IMO this has always been confusing, but IMO in this case can just use "the load test has been executed" or some other variant of the "test was executed" instead of specyfing some code that has run and name it something,

Author: Maybe I'm wrong, but to me this sentence is a key detail:

> Fundamentally, `handleSummary()` is just a function that can access a data object.

So, I'm mildly against removing the reference to `object` in other places because:
- It causes the doc to lose coherence (unless we remove all mentions of the data object).
- It's just less specific, calling it a "summary" doesn't describe the data as well as calling it a "summary object". And working with objects is a fundamental aspect of JavaScript programming, so I don't think it adds new vocabulary or unnecessary complication. Am i making an incorrect assumption here?

On the other hand, I agree with removing the reference to VU code and to callbacks, because these internal details don't really inform how a developer would code a `handleSummary()`.

Reviewer 2: > So, I'm mildly against removing the reference to object in other places because:

I am also against not calling it `object` or `data` :thinking:. Object is bit vague in general ... so maybe `JavaScript object` just to make certain people understand that in this case we do mean a simpla javascript object.

> You can use handleSummary() to manipulate the test result summary as you want. 

IMO "manimulate" is the wrong word. You return the summary you want, you don't manipulate anything. In practice you can return w/e you want even something completely unrelated to the summary object provided.

The `handleSummary()` function takes this object as an argument (called `data` in all examples here).

> k6 calls `handleSummary()` at the end of the test run, even after [`teardown()`](/using-k6/test-life-cycle).

Reviewer: > k6 calls `handleSummary()` at the end of the test run, even after [`teardown()`](/using-k6/test-life-cycle).

I think this could be explained later.

> You can use `handleSummary()` to manipulate this object as you want.

Author: ```suggestion
You can use `handleSummary()` to create a custom summary or return the default summary object.
```

To get an idea of what the data looks like,

> run this script and open the output file, `summary.json`.

Reviewer 2: I do remember there were some differences betweeen this and the file you get from the [summary export](https://k6.io/docs/using-k6/k6-options/reference/#summary-export). 

But I guess we should say that somewhere and also specify you can skip this whole thing by using it. :thinking:

<CodeGroup labels={["return summary as JSON"]} lineNumbers={[]} showCopyButton={[true]}>

```javascript
import http from 'k6/http';

export default function () {
  http.get('https://test.k6.io');
}

export function handleSummary(data) {
  return {
    'summary.json': JSON.stringify(data), //the default data object
  };
}
```

</CodeGroup>

Fundamentally, `handleSummary()` is just a function that can access a data object.
As such, you can transform the summary data into any text format: JSON, HTML, console, XML, and so on.

> You can pipe your custom summary to standard out or standard error, write it to a file, or send it to a remote server.

Author: ```suggestion
You can pipe your custom summary to [standard output or standard error](https://en.wikipedia.org/wiki/Standard_streams), write it to a file, or send it to a remote server.
```

## Use handleSummary()

The following sections go over the `handleSummary()` syntax and provide some examples.

To look up the structure of the summary object, refer to the reference section.

### Syntax

k6 expects `handleSummary()` to return a `{key1: value1, key2: value2, ...}` map that represents the summary metrics.

The keys must be strings.
They determine where k6 displays or saves the content:

> - `stdout` for [standard output](https://en.wikipedia.org/wiki/Standard_streams)
> - `stderr` for standard error,
> - any relative or absolute path to a file on the system (this operation overwrites existing files)

Author: ```suggestion
- `stdout` for standard output
- `stderr` for standard error,
- any relative or absolute path to a file on the system (this operation overwrites existing files)
```

The value of a key can have a type of either `string` or [`ArrayBuffer`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer).

You can return mutiple summary outputs in a script.

> As an example, this `return` statement sends a report to standard out and writes the `data` object to a JSON file.

Author: ```suggestion
As an example, this `return` statement sends a report to standard output and writes the `data` object to a JSON file.
```

</CodeGroup>

> ### Extract data properties

Reviewer: Suggest making it explicit that this is a (dummy) example.

### Example: extract data properties

??

> This minimal `handleSummary()` extracts the `median` value for the `iteration_duration` metric and prints it on standard out:

Author: ```suggestion
This minimal `handleSummary()` extracts the `median` value for the `iteration_duration` metric and prints it to standard output:
```

<CodeGroup labels={["Print metric value"]} lineNumbers={[]} showCopyButton={[true]}>

```javascript
import http from 'k6/http';

export default function () {
  http.get('https://test.k6.io');
}

export function handleSummary(data) {
  const med_latency = data.metrics.iteration_duration.values.med;
  const latency_message = `The median latency was ${med_latency}\n`;

  return {
    stdout: latency_message,
  };
}
```

</CodeGroup>

> ### Modify default output

Reviewer: Same here:

Example: modify default output

     iterations.....................: 5     3.352646/s
     vus............................: 1     min=1      max=1
     vus_max........................: 1     min=1      max=1
```

```
→    data_received..............: 63 kB 39 kB/s
→    data_sent..................: 830 B 507 B/s
→    http_req_blocked...........: med=10.98µs  count=5 p(99)=485.16ms p(99.99)=505.18ms
→    http_req_connecting........: med=0s       count=5 p(99)=245.05ms p(99.99)=255.15ms
→    http_req_duration..........: med=208.68ms count=5 p(99)=302.87ms p(99.99)=306.61ms
→    http_req_failed............: 0.00% ✓ 0        ✗ 5
→    http_req_receiving.........: med=206.12µs count=5 p(99)=341.05µs p(99.99)=344.13µs
→    http_req_sending...........: med=47.8µs   count=5 p(99)=166.94µs p(99.99)=170.92µs
→    http_req_tls_handshaking...: med=0s       count=5 p(99)=207.2ms  p(99.99)=215.74ms
→    http_req_waiting...........: med=208.47ms count=5 p(99)=302.49ms p(99.99)=306.23ms
→    http_reqs..................: 5     3.054928/s
→    vus........................: 1     min=1      max=1
→    vus_max....................: 1     min=1      max=1
```

</CodeGroup>

</Collapsible>

### Custom reports to a remote server

> You can also send the generated reports to a remote server (over any protocol that k6 supports).

Reviewer: 1 - Suggest the same:

- ### Example: custom reports to a remote server

2 - The example says `custom reports to a remote server` but it export to 3 outputs with different formats.

I suggest changing this example to showcase only the output to the remote server. 

If not, the title should be different.
